### ISSUE TITLE
KubernetesPodOperator breaks with active log-collection for long running tasks

### ISSUE THREAD

**Issue by yamrzou** [apache/airflow#12136]
I'm encountering the same bug reported in https://issues.apache.org/jira/browse/AIRFLOW-3534, with airflow 1.10.12.

```
[2020-11-06 13:03:29,672] {pod_launcher.py:173} INFO - Event: fetcher-56104d81c54946a88ce3cd1cf4273477 had an event of type Pending
[2020-11-06 13:03:29,673] {pod_launcher.py:139} WARNING - Pod not yet started: fetcher-56104d81c54946a88ce3cd1cf4273477
[2020-11-06 13:03:30,681] {pod_launcher.py:173} INFO - Event: fetcher-56104d81c54946a88ce3cd1cf4273477 had an event of type Pending
[2020-11-06 13:03:30,681] {pod_launcher.py:139} WARNING - Pod not yet started: fetcher-56104d81c54946a88ce3cd1cf4273477
[2020-11-06 13:03:31,692] {pod_launcher.py:173} INFO - Event: fetcher-56104d81c54946a88ce3cd1cf4273477 had an event of type Pending
[2020-11-06 13:03:31,692] {pod_launcher.py:139} WARNING - Pod not yet started: fetcher-56104d81c54946a88ce3cd1cf4273477
[2020-11-06 13:03:32,702] {pod_launcher.py:173} INFO - Event: fetcher-56104d81c54946a88ce3cd1cf4273477 had an event of type Running
[2020-11-06 13:04:32,740] {taskinstance.py:1150} ERROR - ('Connection broken: IncompleteRead(0 bytes read)', IncompleteRead(0 bytes read))
Traceback (most recent call last):
  File "/opt/bitnami/airflow/venv/lib/python3.6/site-packages/urllib3/response.py", line 696, in _update_chunk_length
    self.chunk_left = int(line, 16)
ValueError: invalid literal for int() with base 16: b''

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/opt/bitnami/airflow/venv/lib/python3.6/site-packages/urllib3/response.py", line 436, in _error_catcher
    yield
  File "/opt/bitnami/airflow/venv/lib/python3.6/site-packages/urllib3/response.py", line 763, in read_chunked
    self._update_chunk_length()
  File "/opt/bitnami/airflow/venv/lib/python3.6/site-packages/urllib3/response.py", line 700, in _update_chunk_length
    raise httplib.IncompleteRead(line)
http.client.IncompleteRead: IncompleteRead(0 bytes read)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/opt/bitnami/airflow/venv/lib/python3.6/site-packages/airflow/models/taskinstance.py", line 979, in _run_raw_task
    result = task_copy.execute(context=context)
  File "/opt/bitnami/airflow/venv/lib/python3.6/site-packages/airflow/contrib/operators/kubernetes_pod_operator.py", line 284, in execute
    final_state, _, result = self.create_new_pod_for_operator(labels, launcher)
  File "/opt/bitnami/airflow/venv/lib/python3.6/site-packages/airflow/contrib/operators/kubernetes_pod_operator.py", line 403, in create_new_pod_for_operator
    final_state, result = launcher.monitor_pod(pod=pod, get_logs=self.get_logs)
  File "/opt/bitnami/airflow/venv/lib/python3.6/site-packages/airflow/kubernetes/pod_launcher.py", line 155, in monitor_pod
    for line in logs:
  File "/opt/bitnami/airflow/venv/lib/python3.6/site-packages/urllib3/response.py", line 807, in __iter__
    for chunk in self.stream(decode_content=True):
  File "/opt/bitnami/airflow/venv/lib/python3.6/site-packages/urllib3/response.py", line 571, in stream
    for line in self.read_chunked(amt, decode_content=decode_content):
  File "/opt/bitnami/airflow/venv/lib/python3.6/site-packages/urllib3/response.py", line 792, in read_chunked
    self._original_response.close()
  File "/opt/bitnami/python/lib/python3.6/contextlib.py", line 99, in __exit__
    self.gen.throw(type, value, traceback)
  File "/opt/bitnami/airflow/venv/lib/python3.6/site-packages/urllib3/response.py", line 454, in _error_catcher
    raise ProtocolError("Connection broken: %r" % e, e)
urllib3.exceptions.ProtocolError: ('Connection broken: IncompleteRead(0 bytes read)', IncompleteRead(0 bytes read))
[2020-11-06 13:04:32,743] {taskinstance.py:1194} INFO - Marking task as UP_FOR_RETRY. dag_id=..., task_id=..., execution_date=20201106T120000, start_date=20201106T130329, end_date=20201106T130432
[2020-11-06 13:04:34,641] {local_task_job.py:102} INFO - Task exited with return code 1
```

The bug goes away by setting `get_logs=False` in the KubernetesPodOperator. Reproduced with multiple dags and tasks.

**Comment by dmateusp** [comment]
hey @yamrzou I know it's been some time since you originally opened that issue! We run Airflow 1.10.10 and noticed the same, we have been getting around it by modifying the containers to frequently log.

Now I'm trying to actually fix the issue but I'm having a hard time reproducing it with a simple task:

```python
    bash_echo_1 = KubernetesPodOperator(
        task_id="bash_echo_1",
        image="bash:4.4",
        name="bash-echo-1",
        cmds=[
            "bash",
            "-c",
            "while true; do sleep 5; done;",
        ],
        do_xcom_push=True,
        get_logs=True,
    )
``` 

Do you have a simple example that reproduces it on your end?

**Comment by dmateusp** [comment]
I was able to reproduce this issue after a few days of letting Bash containers run from KubernetesPodOperator, and implement a fix:

I ran two bash tasks, one with the fix (EarnestKubernetesPodOperator) and one without (KubernetesPodOperator):

```python
    bash_echo_1 = KubernetesPodOperator(
        task_id="bash_echo_1",
        image="bash:4.4",
        name="bash-echo-1",
        cmds=[
            "bash",
            "-c",
            "while true; do sleep 5; done;",
        ],
        do_xcom_push=True,
        get_logs=True,
        **t.DEFAULT_POD_ARGS,
    )

    bash_echo_2 = EarnestKubernetesPodOperator(
        task_id="bash_echo_2",
        name="bash-echo-2",
        image="bash:4.4",
        cmds=["bash", "-c", "while true; do sleep 5; done;",],
        get_logs=True,
        **t.DEFAULT_POD_ARGS,
    )
```

I let both of these run for a few days and the one without the fix failed a few times while the one with the fix never failed:

<img width="95" alt="Screen Shot 2021-01-04 at 14 07 50" src="https://user-images.githubusercontent.com/8832050/103543765-f3f65480-4e96-11eb-9b7d-05f71a9f1036.png">

^ First row is `bash_echo_1` and second row is `bash_echo_2`

In the logs of `bash_echo_1` we can see that the issue was reproduced after 3 hours of running:

```
[2020-12-31 03:00:36,636] {logging_mixin.py:112} INFO - [2020-12-31 03:00:36,635] {pod_launcher.py:142} INFO - Event: bash-echo-1-3a123499 had an event of type Running
[2021-01-01 06:12:42,517] {taskinstance.py:1145} ERROR - ('Connection broken: IncompleteRead(0 bytes read)', IncompleteRead(0 bytes read))
Traceback (most recent call last):
  File "/home/airflow/.local/lib/python3.7/site-packages/urllib3/response.py", line 685, in _update_chunk_length
    self.chunk_left = int(line, 16)
ValueError: invalid literal for int() with base 16: b''

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/airflow/.local/lib/python3.7/site-packages/urllib3/response.py", line 425, in _error_catcher
    yield
  File "/home/airflow/.local/lib/python3.7/site-packages/urllib3/response.py", line 752, in read_chunked
    self._update_chunk_length()
  File "/home/airflow/.local/lib/python3.7/site-packages/urllib3/response.py", line 689, in _update_chunk_length
    raise httplib.IncompleteRead(line)
http.client.IncompleteRead: IncompleteRead(0 bytes read)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/airflow/.local/lib/python3.7/site-packages/airflow/models/taskinstance.py", line 983, in _run_raw_task
    result = task_copy.execute(context=context)
  File "/home/airflow/.local/lib/python3.7/site-packages/airflow/contrib/operators/kubernetes_pod_operator.py", line 251, in execute
    get_logs=self.get_logs)
  File "/home/airflow/.local/lib/python3.7/site-packages/airflow/contrib/kubernetes/pod_launcher.py", line 117, in run_pod
    return self._monitor_pod(pod, get_logs)
  File "/home/airflow/.local/lib/python3.7/site-packages/airflow/contrib/kubernetes/pod_launcher.py", line 124, in _monitor_pod
    for line in logs:
  File "/home/airflow/.local/lib/python3.7/site-packages/urllib3/response.py", line 796, in __iter__
    for chunk in self.stream(decode_content=True):
  File "/home/airflow/.local/lib/python3.7/site-packages/urllib3/response.py", line 560, in stream
    for line in self.read_chunked(amt, decode_content=decode_content):
  File "/home/airflow/.local/lib/python3.7/site-packages/urllib3/response.py", line 781, in read_chunked
    self._original_response.close()
  File "/usr/local/lib/python3.7/contextlib.py", line 130, in __exit__
    self.gen.throw(type, value, traceback)
  File "/home/airflow/.local/lib/python3.7/site-packages/urllib3/response.py", line 443, in _error_catcher
    raise ProtocolError("Connection broken: %r" % e, e)
urllib3.exceptions.ProtocolError: ('Connection broken: IncompleteRead(0 bytes read)', IncompleteRead(0 bytes read))
```

While the same issue happened in `bash_echo_2` but it was handled by the change in this PR

```
[2020-12-31 03:00:36,592] {logging_mixin.py:112} INFO - [2020-12-31 03:00:36,592] {kubernetes_pod_operator.py:270} INFO - Event: bash-echo-2-72839c01adc84a29b3903527d6e6dfba had an event of type Running
[2021-01-01 06:20:38,083] {logging_mixin.py:112} INFO - [2021-01-01 06:20:38,082] {kubernetes_pod_operator.py:248} INFO - The pod has not logged since the logs were last fetched
[2021-01-01 06:20:38,373] {logging_mixin.py:112} INFO - [2021-01-01 06:20:38,372] {kubernetes_pod_operator.py:270} INFO - Event: bash-echo-2-72839c01adc84a29b3903527d6e6dfba had an event of type Running
```

The only change was to add error handling around `self.read_pod_logs` in the KubernetesPodOperator:

```python
            try:
                logs = self.read_pod_logs(pod)
                for line in logs:
                    self.log.info(line)
            except ProtocolError as pe:
                _, protocol_exception = pe.args
                # When no logs are fetched, an IncompleteRead is thrown trying
                # to decode the stream
                if str(protocol_exception) == "IncompleteRead(0 bytes read)":
                    self.log.info(
                        "The pod has not logged since the logs were last fetched"
                    )
                # If the exception is not about an empty stream we raise it
                else:
                    raise e
```

**Comment by yamrzou** [comment]
Hello @dmateusp ,

Thank you for investigating and for sharing your fix!

Regarding the reproducibility of the issue, I noticed that I can not reproduce it when using the following `cmd` and `args`:
```
cmds=["bash", "-c"],
arguments=['sleep 60']
Or 
arguments=['echo hello && sleep 60']
```
But I systematically reproduce it with:
```
cmds=["bash", "-c"],
arguments=['python -c "print(\\"hello\\")" && sleep 60']
```

I tested your proposed solution and it worked. So until a PR and a new version is released, I'm monkey-patching the `PodLauncher` class within the dag code as follows:

```
from airflow.kubernetes.pod_launcher import PodLauncher
from urllib3.exceptions import ProtocolError

def patch_read_pod_logs(func):
    def wrapper(self, *args, **kwargs):
        logs = func(self, *args, **kwargs)
        try:
            for line in logs:
                yield line
        except ProtocolError as pe:
            _, protocol_exception = pe.args
            # When no logs are fetched, an IncompleteRead is thrown trying
            # to decode the stream
            if str(protocol_exception) == "IncompleteRead(0 bytes read)":
                self.log.info(
                    "The pod has not logged since the logs were last fetched"
                )
            else:
                # If the exception is not about an empty stream we raise it
                raise pe
    return wrapper

PodLauncher.read_pod_logs = patch_read_pod_logs(PodLauncher.read_pod_logs)
```

**Comment by yamrzou** [comment]
@dmateusp I found an issue with the above fix. When running with the followings arguments, only the first `print()` is logged, the later ones are discarded:
```
cmds=["bash", "-c"],
arguments=[
    'python -c "print(\\"hello\\")" && sleep 120 && python -c "print(\\"hello again\\")" && sleep 30 && python -c "print(\\"last hello\\")" && sleep 30'
]
```
Log output:
```
[2021-01-07 16:20:55,327] {pod_launcher.py:173} INFO - Event: test-2a1b09d88b9340a291751b9c1f536d7b had an event of type Running
[2021-01-07 16:21:55,331] {pod_launcher.py:156} INFO - b'hello\n'
[2021-01-07 16:21:55,332] {sample_dag.py:26} INFO - The pod has not logged since the logs were last fetched
[2021-01-07 16:21:55,365] {pod_launcher.py:173} INFO - Event: test-2a1b09d88b9340a291751b9c1f536d7b had an event of type Running
[2021-01-07 16:21:55,365] {pod_launcher.py:166} INFO - Pod test-2a1b09d88b9340a291751b9c1f536d7b has state running
[2021-01-07 16:21:57,374] {pod_launcher.py:173} INFO - Event: test-2a1b09d88b9340a291751b9c1f536d7b had an event of type Running
[2021-01-07 16:21:57,375] {pod_launcher.py:166} INFO - Pod test-2a1b09d88b9340a291751b9c1f536d7b has state running
[2021-01-07 16:21:59,385] {pod_launcher.py:173} INFO - Event: test-2a1b09d88b9340a291751b9c1f536d7b had an event of type Running
[2021-01-07 16:21:59,385] {pod_launcher.py:166} INFO - Pod test-2a1b09d88b9340a291751b9c1f536d7b has state running
[2021-01-07 16:22:01,398] {pod_launcher.py:173} INFO - Event: test-2a1b09d88b9340a291751b9c1f536d7b had an event of type Running
[2021-01-07 16:22:01,398] {pod_launcher.py:166} INFO - Pod test-2a1b09d88b9340a291751b9c1f536d7b has state running
[2021-01-07 16:22:03,409] {pod_launcher.py:173} INFO - Event: test-2a1b09d88b9340a291751b9c1f536d7b had an event of type Running
[2021-01-07 16:22:03,409] {pod_launcher.py:166} INFO - Pod test-2a1b09d88b9340a291751b9c1f536d7b has state running
[2021-01-07 16:22:05,420] {pod_launcher.py:173} INFO - Event: test-2a1b09d88b9340a291751b9c1f536d7b had an event of type Running
[2021-01-07 16:22:05,420] {pod_launcher.py:166} INFO - Pod test-2a1b09d88b9340a291751b9c1f536d7b has state running
[2021-01-07 16:22:07,435] {pod_launcher.py:173} INFO - Event: test-2a1b09d88b9340a291751b9c1f536d7b had an event of type Running
[2021-01-07 16:22:07,436] {pod_launcher.py:166} INFO - Pod test-2a1b09d88b9340a291751b9c1f536d7b has state running
[2021-01-07 16:22:09,451] {pod_launcher.py:173} INFO - Event: test-2a1b09d88b9340a291751b9c1f536d7b had an event of type Running
[2021-01-07 16:22:09,452] {pod_launcher.py:166} INFO - Pod test-2a1b09d88b9340a291751b9c1f536d7b has state running
[2021-01-07 16:22:11,463] {pod_launcher.py:173} INFO - Event: test-2a1b09d88b9340a291751b9c1f536d7b had an event of type Running
[2021-01-07 16:22:11,463] {pod_launcher.py:166} INFO - Pod test-2a1b09d88b9340a291751b9c1f536d7b has state running
[2021-01-07 16:22:13,474] {pod_launcher.py:173} INFO - Event: test-2a1b09d88b9340a291751b9c1f536d7b had an event of type Running
[2021-01-07 16:22:13,474] {pod_launcher.py:166} INFO - Pod test-2a1b09d88b9340a291751b9c1f536d7b has state running
[2021-01-07 16:22:15,495] {pod_launcher.py:173} INFO - Event: test-2a1b09d88b9340a291751b9c1f536d7b had an event of type Running
[2021-01-07 16:22:15,495] {pod_launcher.py:166} INFO - Pod test-2a1b09d88b9340a291751b9c1f536d7b has state running
[2021-01-07 16:22:17,506] {pod_launcher.py:173} INFO - Event: test-2a1b09d88b9340a291751b9c1f536d7b had an event of type Running
[2021-01-07 16:22:17,506] {pod_launcher.py:166} INFO - Pod test-2a1b09d88b9340a291751b9c1f536d7b has state running
[2021-01-07 16:22:19,527] {pod_launcher.py:173} INFO - Event: test-2a1b09d88b9340a291751b9c1f536d7b had an event of type Running
[2021-01-07 16:22:19,527] {pod_launcher.py:166} INFO - Pod test-2a1b09d88b9340a291751b9c1f536d7b has state running
[2021-01-07 16:22:21,538] {pod_launcher.py:173} INFO - Event: test-2a1b09d88b9340a291751b9c1f536d7b had an event of type Running
[2021-01-07 16:22:21,538] {pod_launcher.py:166} INFO - Pod test-2a1b09d88b9340a291751b9c1f536d7b has state running
[2021-01-07 16:22:23,549] {pod_launcher.py:173} INFO - Event: test-2a1b09d88b9340a291751b9c1f536d7b had an event of type Running
[2021-01-07 16:22:23,549] {pod_launcher.py:166} INFO - Pod test-2a1b09d88b9340a291751b9c1f536d7b has state running
[2021-01-07 16:22:25,559] {pod_launcher.py:173} INFO - Event: test-2a1b09d88b9340a291751b9c1f536d7b had an event of type Running
[2021-01-07 16:22:25,559] {pod_launcher.py:166} INFO - Pod test-2a1b09d88b9340a291751b9c1f536d7b has state running
[2021-01-07 16:22:27,572] {pod_launcher.py:173} INFO - Event: test-2a1b09d88b9340a291751b9c1f536d7b had an event of type Running
[2021-01-07 16:22:27,572] {pod_launcher.py:166} INFO - Pod test-2a1b09d88b9340a291751b9c1f536d7b has state running
[2021-01-07 16:22:29,588] {pod_launcher.py:173} INFO - Event: test-2a1b09d88b9340a291751b9c1f536d7b had an event of type Running
[2021-01-07 16:22:29,588] {pod_launcher.py:166} INFO - Pod test-2a1b09d88b9340a291751b9c1f536d7b has state running
[2021-01-07 16:22:31,598] {pod_launcher.py:173} INFO - Event: test-2a1b09d88b9340a291751b9c1f536d7b had an event of type Running
[2021-01-07 16:22:31,598] {pod_launcher.py:166} INFO - Pod test-2a1b09d88b9340a291751b9c1f536d7b has state running
[2021-01-07 16:22:33,608] {pod_launcher.py:173} INFO - Event: test-2a1b09d88b9340a291751b9c1f536d7b had an event of type Running
[2021-01-07 16:22:33,608] {pod_launcher.py:166} INFO - Pod test-2a1b09d88b9340a291751b9c1f536d7b has state running
[2021-01-07 16:22:35,619] {pod_launcher.py:173} INFO - Event: test-2a1b09d88b9340a291751b9c1f536d7b had an event of type Running
[2021-01-07 16:22:35,619] {pod_launcher.py:166} INFO - Pod test-2a1b09d88b9340a291751b9c1f536d7b has state running
[2021-01-07 16:22:37,630] {pod_launcher.py:173} INFO - Event: test-2a1b09d88b9340a291751b9c1f536d7b had an event of type Running
[2021-01-07 16:22:37,630] {pod_launcher.py:166} INFO - Pod test-2a1b09d88b9340a291751b9c1f536d7b has state running
[2021-01-07 16:22:39,641] {pod_launcher.py:173} INFO - Event: test-2a1b09d88b9340a291751b9c1f536d7b had an event of type Running
[2021-01-07 16:22:39,642] {pod_launcher.py:166} INFO - Pod test-2a1b09d88b9340a291751b9c1f536d7b has state running
[2021-01-07 16:22:41,650] {pod_launcher.py:173} INFO - Event: test-2a1b09d88b9340a291751b9c1f536d7b had an event of type Running
[2021-01-07 16:22:41,650] {pod_launcher.py:166} INFO - Pod test-2a1b09d88b9340a291751b9c1f536d7b has state running
[2021-01-07 16:22:43,662] {pod_launcher.py:173} INFO - Event: test-2a1b09d88b9340a291751b9c1f536d7b had an event of type Running
[2021-01-07 16:22:43,662] {pod_launcher.py:166} INFO - Pod test-2a1b09d88b9340a291751b9c1f536d7b has state running
[2021-01-07 16:22:45,672] {pod_launcher.py:173} INFO - Event: test-2a1b09d88b9340a291751b9c1f536d7b had an event of type Running
[2021-01-07 16:22:45,672] {pod_launcher.py:166} INFO - Pod test-2a1b09d88b9340a291751b9c1f536d7b has state running
[2021-01-07 16:22:47,683] {pod_launcher.py:173} INFO - Event: test-2a1b09d88b9340a291751b9c1f536d7b had an event of type Running
[2021-01-07 16:22:47,683] {pod_launcher.py:166} INFO - Pod test-2a1b09d88b9340a291751b9c1f536d7b has state running
[2021-01-07 16:22:49,693] {pod_launcher.py:173} INFO - Event: test-2a1b09d88b9340a291751b9c1f536d7b had an event of type Running
[2021-01-07 16:22:49,693] {pod_launcher.py:166} INFO - Pod test-2a1b09d88b9340a291751b9c1f536d7b has state running
[2021-01-07 16:22:51,707] {pod_launcher.py:173} INFO - Event: test-2a1b09d88b9340a291751b9c1f536d7b had an event of type Running
[2021-01-07 16:22:51,707] {pod_launcher.py:166} INFO - Pod test-2a1b09d88b9340a291751b9c1f536d7b has state running
[2021-01-07 16:22:53,719] {pod_launcher.py:173} INFO - Event: test-2a1b09d88b9340a291751b9c1f536d7b had an event of type Running
[2021-01-07 16:22:53,719] {pod_launcher.py:166} INFO - Pod test-2a1b09d88b9340a291751b9c1f536d7b has state running
[2021-01-07 16:22:55,728] {pod_launcher.py:173} INFO - Event: test-2a1b09d88b9340a291751b9c1f536d7b had an event of type Running
[2021-01-07 16:22:55,729] {pod_launcher.py:166} INFO - Pod test-2a1b09d88b9340a291751b9c1f536d7b has state running
[2021-01-07 16:22:57,740] {pod_launcher.py:173} INFO - Event: test-2a1b09d88b9340a291751b9c1f536d7b had an event of type Running
[2021-01-07 16:22:57,740] {pod_launcher.py:166} INFO - Pod test-2a1b09d88b9340a291751b9c1f536d7b has state running
[2021-01-07 16:22:59,751] {pod_launcher.py:173} INFO - Event: test-2a1b09d88b9340a291751b9c1f536d7b had an event of type Running
[2021-01-07 16:22:59,751] {pod_launcher.py:166} INFO - Pod test-2a1b09d88b9340a291751b9c1f536d7b has state running
[2021-01-07 16:23:01,766] {pod_launcher.py:173} INFO - Event: test-2a1b09d88b9340a291751b9c1f536d7b had an event of type Running
[2021-01-07 16:23:01,766] {pod_launcher.py:166} INFO - Pod test-2a1b09d88b9340a291751b9c1f536d7b has state running
[2021-01-07 16:23:03,776] {pod_launcher.py:173} INFO - Event: test-2a1b09d88b9340a291751b9c1f536d7b had an event of type Running
[2021-01-07 16:23:03,776] {pod_launcher.py:166} INFO - Pod test-2a1b09d88b9340a291751b9c1f536d7b has state running
[2021-01-07 16:23:05,794] {pod_launcher.py:173} INFO - Event: test-2a1b09d88b9340a291751b9c1f536d7b had an event of type Running
[2021-01-07 16:23:05,794] {pod_launcher.py:166} INFO - Pod test-2a1b09d88b9340a291751b9c1f536d7b has state running
[2021-01-07 16:23:07,807] {pod_launcher.py:173} INFO - Event: test-2a1b09d88b9340a291751b9c1f536d7b had an event of type Running
[2021-01-07 16:23:07,808] {pod_launcher.py:166} INFO - Pod test-2a1b09d88b9340a291751b9c1f536d7b has state running
[2021-01-07 16:23:09,819] {pod_launcher.py:173} INFO - Event: test-2a1b09d88b9340a291751b9c1f536d7b had an event of type Running
[2021-01-07 16:23:09,820] {pod_launcher.py:166} INFO - Pod test-2a1b09d88b9340a291751b9c1f536d7b has state running
[2021-01-07 16:23:11,832] {pod_launcher.py:173} INFO - Event: test-2a1b09d88b9340a291751b9c1f536d7b had an event of type Running
[2021-01-07 16:23:11,832] {pod_launcher.py:166} INFO - Pod test-2a1b09d88b9340a291751b9c1f536d7b has state running
[2021-01-07 16:23:13,846] {pod_launcher.py:173} INFO - Event: test-2a1b09d88b9340a291751b9c1f536d7b had an event of type Running
[2021-01-07 16:23:13,846] {pod_launcher.py:166} INFO - Pod test-2a1b09d88b9340a291751b9c1f536d7b has state running
[2021-01-07 16:23:15,858] {pod_launcher.py:173} INFO - Event: test-2a1b09d88b9340a291751b9c1f536d7b had an event of type Running
[2021-01-07 16:23:15,858] {pod_launcher.py:166} INFO - Pod test-2a1b09d88b9340a291751b9c1f536d7b has state running
[2021-01-07 16:23:17,872] {pod_launcher.py:173} INFO - Event: test-2a1b09d88b9340a291751b9c1f536d7b had an event of type Running
[2021-01-07 16:23:17,872] {pod_launcher.py:166} INFO - Pod test-2a1b09d88b9340a291751b9c1f536d7b has state running
[2021-01-07 16:23:19,884] {pod_launcher.py:173} INFO - Event: test-2a1b09d88b9340a291751b9c1f536d7b had an event of type Running
[2021-01-07 16:23:19,885] {pod_launcher.py:166} INFO - Pod test-2a1b09d88b9340a291751b9c1f536d7b has state running
[2021-01-07 16:23:21,897] {pod_launcher.py:173} INFO - Event: test-2a1b09d88b9340a291751b9c1f536d7b had an event of type Running
[2021-01-07 16:23:21,897] {pod_launcher.py:166} INFO - Pod test-2a1b09d88b9340a291751b9c1f536d7b has state running
[2021-01-07 16:23:23,907] {pod_launcher.py:173} INFO - Event: test-2a1b09d88b9340a291751b9c1f536d7b had an event of type Running
[2021-01-07 16:23:23,907] {pod_launcher.py:166} INFO - Pod test-2a1b09d88b9340a291751b9c1f536d7b has state running
[2021-01-07 16:23:25,919] {pod_launcher.py:173} INFO - Event: test-2a1b09d88b9340a291751b9c1f536d7b had an event of type Running
[2021-01-07 16:23:25,920] {pod_launcher.py:166} INFO - Pod test-2a1b09d88b9340a291751b9c1f536d7b has state running
[2021-01-07 16:23:27,928] {pod_launcher.py:173} INFO - Event: test-2a1b09d88b9340a291751b9c1f536d7b had an event of type Running
[2021-01-07 16:23:27,928] {pod_launcher.py:166} INFO - Pod test-2a1b09d88b9340a291751b9c1f536d7b has state running
[2021-01-07 16:23:29,941] {pod_launcher.py:173} INFO - Event: test-2a1b09d88b9340a291751b9c1f536d7b had an event of type Running
[2021-01-07 16:23:29,941] {pod_launcher.py:166} INFO - Pod test-2a1b09d88b9340a291751b9c1f536d7b has state running
[2021-01-07 16:23:31,953] {pod_launcher.py:173} INFO - Event: test-2a1b09d88b9340a291751b9c1f536d7b had an event of type Running
[2021-01-07 16:23:31,953] {pod_launcher.py:166} INFO - Pod test-2a1b09d88b9340a291751b9c1f536d7b has state running
[2021-01-07 16:23:33,968] {pod_launcher.py:173} INFO - Event: test-2a1b09d88b9340a291751b9c1f536d7b had an event of type Running
[2021-01-07 16:23:33,968] {pod_launcher.py:166} INFO - Pod test-2a1b09d88b9340a291751b9c1f536d7b has state running
[2021-01-07 16:23:35,979] {pod_launcher.py:173} INFO - Event: test-2a1b09d88b9340a291751b9c1f536d7b had an event of type Running
[2021-01-07 16:23:35,979] {pod_launcher.py:166} INFO - Pod test-2a1b09d88b9340a291751b9c1f536d7b has state running
[2021-01-07 16:23:37,991] {pod_launcher.py:173} INFO - Event: test-2a1b09d88b9340a291751b9c1f536d7b had an event of type Running
[2021-01-07 16:23:37,991] {pod_launcher.py:166} INFO - Pod test-2a1b09d88b9340a291751b9c1f536d7b has state running
[2021-01-07 16:23:40,002] {pod_launcher.py:173} INFO - Event: test-2a1b09d88b9340a291751b9c1f536d7b had an event of type Running
[2021-01-07 16:23:40,002] {pod_launcher.py:166} INFO - Pod test-2a1b09d88b9340a291751b9c1f536d7b has state running
[2021-01-07 16:23:42,014] {pod_launcher.py:173} INFO - Event: test-2a1b09d88b9340a291751b9c1f536d7b had an event of type Running
[2021-01-07 16:23:42,014] {pod_launcher.py:166} INFO - Pod test-2a1b09d88b9340a291751b9c1f536d7b has state running
[2021-01-07 16:23:44,026] {pod_launcher.py:173} INFO - Event: test-2a1b09d88b9340a291751b9c1f536d7b had an event of type Running
[2021-01-07 16:23:44,026] {pod_launcher.py:166} INFO - Pod test-2a1b09d88b9340a291751b9c1f536d7b has state running
[2021-01-07 16:23:46,039] {pod_launcher.py:173} INFO - Event: test-2a1b09d88b9340a291751b9c1f536d7b had an event of type Running
[2021-01-07 16:23:46,039] {pod_launcher.py:166} INFO - Pod test-2a1b09d88b9340a291751b9c1f536d7b has state running
[2021-01-07 16:23:48,050] {pod_launcher.py:173} INFO - Event: test-2a1b09d88b9340a291751b9c1f536d7b had an event of type Running
[2021-01-07 16:23:48,051] {pod_launcher.py:166} INFO - Pod test-2a1b09d88b9340a291751b9c1f536d7b has state running
[2021-01-07 16:23:50,061] {pod_launcher.py:173} INFO - Event: test-2a1b09d88b9340a291751b9c1f536d7b had an event of type Running
[2021-01-07 16:23:50,061] {pod_launcher.py:166} INFO - Pod test-2a1b09d88b9340a291751b9c1f536d7b has state running
[2021-01-07 16:23:52,071] {pod_launcher.py:173} INFO - Event: test-2a1b09d88b9340a291751b9c1f536d7b had an event of type Running
[2021-01-07 16:23:52,071] {pod_launcher.py:166} INFO - Pod test-2a1b09d88b9340a291751b9c1f536d7b has state running
[2021-01-07 16:23:54,082] {pod_launcher.py:173} INFO - Event: test-2a1b09d88b9340a291751b9c1f536d7b had an event of type Running
[2021-01-07 16:23:54,082] {pod_launcher.py:166} INFO - Pod test-2a1b09d88b9340a291751b9c1f536d7b has state running
[2021-01-07 16:23:56,095] {pod_launcher.py:173} INFO - Event: test-2a1b09d88b9340a291751b9c1f536d7b had an event of type Succeeded
[2021-01-07 16:23:56,095] {pod_launcher.py:287} INFO - Event with job id test-2a1b09d88b9340a291751b9c1f536d7b Succeeded
[2021-01-07 16:23:56,106] {pod_launcher.py:173} INFO - Event: test-2a1b09d88b9340a291751b9c1f536d7b had an event of type Succeeded
[2021-01-07 16:23:56,106] {pod_launcher.py:287} INFO - Event with job id test-2a1b09d88b9340a291751b9c1f536d7b Succeeded
[2021-01-07 16:23:56,158] {taskinstance.py:1070} INFO - Marking task as SUCCESS.dag_id=sample_dag_2, task_id=kubernetes_task_2, execution_date=20210107T162031, start_date=20210107T162051, end_date=20210107T162356
[2021-01-07 16:23:57,928] {local_task_job.py:102} INFO - Task exited with return code 0
```

Expected log output, when using the following arguments (reduced sleep time to avoid `IncompleteRead`): 
``` 
'python -c "print(\\"hello\\")" && sleep 5 && python -c "print(\\"hello again\\")" && sleep 5 && python -c "print(\\"last hello\\")" && sleep 5'
```

```
[2021-01-07 16:35:41,059] {pod_launcher.py:173} INFO - Event: test-3a5e94a9eb344c2786e583024c9ccdfc had an event of type Pending
[2021-01-07 16:35:41,059] {pod_launcher.py:139} WARNING - Pod not yet started: test-3a5e94a9eb344c2786e583024c9ccdfc
[2021-01-07 16:35:42,072] {pod_launcher.py:173} INFO - Event: test-3a5e94a9eb344c2786e583024c9ccdfc had an event of type Running
[2021-01-07 16:35:57,773] {pod_launcher.py:156} INFO - b'hello\n'
[2021-01-07 16:35:57,774] {pod_launcher.py:156} INFO - b'hello again\n'
[2021-01-07 16:35:57,774] {pod_launcher.py:156} INFO - b'last hello\n'
[2021-01-07 16:35:57,790] {pod_launcher.py:173} INFO - Event: test-3a5e94a9eb344c2786e583024c9ccdfc had an event of type Succeeded
[2021-01-07 16:35:57,791] {pod_launcher.py:287} INFO - Event with job id test-3a5e94a9eb344c2786e583024c9ccdfc Succeeded
[2021-01-07 16:35:57,798] {pod_launcher.py:173} INFO - Event: test-3a5e94a9eb344c2786e583024c9ccdfc had an event of type Succeeded
[2021-01-07 16:35:57,798] {pod_launcher.py:287} INFO - Event with job id test-3a5e94a9eb344c2786e583024c9ccdfc Succeeded
[2021-01-07 16:35:57,844] {taskinstance.py:1070} INFO - Marking task as SUCCESS.dag_id=sample_dag_2, task_id=kubernetes_task_2, execution_date=20210107T163523, start_date=20210107T163538, end_date=20210107T163557

```

**Comment by dmateusp** [comment]
hey thanks for reporting the issue, I still don't have a good way to reproduce it unfortunately :( and when I attempted to fix on the master branch, I saw that the relevant part of the code for Airflow 2.0 changed quite a bit.

Ideally I'd have a test against the 1.10 branch that reproduces the issue, but for now I'm going to close my PR to master

**Comment by saisujithkp** [comment]
Is anyone still having the issue mentioned by @yamrzou , I'm using airflow 1.10.14 and have a similar issue where the logs of a long running task are similar to https://github.com/apache/airflow/issues/12136#issuecomment-756233260.

**Comment by ausiddiqui** [comment]
Yes I'm having this issue running Airflow 1.10.12 with KubernetesPodOperator which is running a queries packaged in a Docker image. We are running long-running queries where the queries will be running and producing the logs in Airflow for 2-4 hours and then at 3hrs 59 minutes of query logs we would just get:
```log
INFO - Event <pod_name> had an event of type Running
INFO - Event: <pod_name> has state running
```
 repeated every 2 seconds for 2.5 to 3.5 hours until a failed state.
```log
INFO - Event <pod_name> had an event of type Failed
```

> Is anyone still having the issue mentioned by @yamrzou , I'm using airflow 1.10.14 and have a similar issue where the logs of a long running task are similar to [#12136 (comment)](https://github.com/apache/airflow/issues/12136#issuecomment-756233260).

**Comment by talnicolas** [comment]
I confirm that I'm still having this issue in Airflow 2.0.2.

**Comment by cchicote** [comment]
I have the exact same issue than @ausiddiqui, running an MWAA environment with Airflow 1.10.12, along with an EKS cluster with Kubernetes 1.19
Anyone had such an issue and found a solution ?

**Comment by atejano** [comment]
I also have this issue. I am running Airflow 1.10.15. Pods are still running in kubernetes cluster but airflow seem to be disconnected to the pod after 4hrs.

**Comment by hsnprsd** [comment]
We are having the same issue. Here is the log:
```
[2021-06-19 09:12:08,519] {taskinstance.py:1482} ERROR - Task failed with exception
Traceback (most recent call last):
  File "/home/airflow/.local/lib/python3.6/site-packages/urllib3/response.py", line 696, in _update_chunk_length
    self.chunk_left = int(line, 16)
ValueError: invalid literal for int() with base 16: b''

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/airflow/.local/lib/python3.6/site-packages/urllib3/response.py", line 436, in _error_catcher
    yield
  File "/home/airflow/.local/lib/python3.6/site-packages/urllib3/response.py", line 763, in read_chunked
    self._update_chunk_length()
  File "/home/airflow/.local/lib/python3.6/site-packages/urllib3/response.py", line 700, in _update_chunk_length
    raise httplib.IncompleteRead(line)
http.client.IncompleteRead: IncompleteRead(0 bytes read)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/airflow/.local/lib/python3.6/site-packages/airflow/models/taskinstance.py", line 1138, in _run_raw_task
    self._prepare_and_execute_task_with_callbacks(context, task)
  File "/home/airflow/.local/lib/python3.6/site-packages/airflow/models/taskinstance.py", line 1311, in _prepare_and_execute_task_with_callbacks
    result = self._execute_task(context, task_copy)
  File "/home/airflow/.local/lib/python3.6/site-packages/airflow/models/taskinstance.py", line 1336, in _execute_task
    result = task_copy.execute(context=context)
  File "/home/airflow/.local/lib/python3.6/site-packages/airflow/providers/cncf/kubernetes/operators/kubernetes_pod.py", line 361, in execute
    final_state, _, result = self.create_new_pod_for_operator(labels, launcher)
  File "/home/airflow/.local/lib/python3.6/site-packages/airflow/providers/cncf/kubernetes/operators/kubernetes_pod.py", line 508, in create_new_pod_for_operator
    final_state, result = launcher.monitor_pod(pod=self.pod, get_logs=self.get_logs)
  File "/home/airflow/.local/lib/python3.6/site-packages/airflow/kubernetes/pod_launcher.py", line 133, in monitor_pod
    for line in logs:
  File "/home/airflow/.local/lib/python3.6/site-packages/urllib3/response.py", line 807, in __iter__
    for chunk in self.stream(decode_content=True):
  File "/home/airflow/.local/lib/python3.6/site-packages/urllib3/response.py", line 571, in stream
    for line in self.read_chunked(amt, decode_content=decode_content):
  File "/home/airflow/.local/lib/python3.6/site-packages/urllib3/response.py", line 792, in read_chunked
    self._original_response.close()
  File "/usr/local/lib/python3.6/contextlib.py", line 99, in __exit__
    self.gen.throw(type, value, traceback)
  File "/home/airflow/.local/lib/python3.6/site-packages/urllib3/response.py", line 454, in _error_catcher
    raise ProtocolError("Connection broken: %r" % e, e)
urllib3.exceptions.ProtocolError: ('Connection broken: IncompleteRead(0 bytes read)', IncompleteRead(0 bytes read))
```

**Comment by cchicote** [comment]
Hi everyone,

Updating my MWAA environment from Airflow 1.10.12 to Airflow 2.0.2 along with the configuration of log levels to INFO for the task and the worker on the MWAA environment allowed me to get the logs of my tasks on the Airflow UI, and to keep getting them after 4 hours ! (issue that @ausiddiqui encountered)

To test it, i configured this pod that i let run for a little bit more than 24h:
```
podRun = KubernetesPodOperator(
                       namespace="mwaa",
                       image="ubuntu:18.04",
                       cmds=["bash"],
                       arguments=["-c", "while true; do date; sleep 10; done"],
                       labels={"foo": "bar"},
                       name="mwaa-pod-test",
                       task_id="pod-task",
                       get_logs=True,
                       dag=dag,
                       is_delete_operator_pod=False,
                       config_file=kube_config_path,
                       in_cluster=False,
                       cluster_context='aws'
                       )
```

I noticed that there is a warning message after 4 hours of execution, and a duplication of the last message received before the warning (looks like it's disconnecting, reconnecting and then sending the last log received), this happens every 4 hours, until the end of the execution of the task:
```
[2021-06-23 12:40:08,015] {{logging_mixin.py:104}} INFO - [2021-06-23 12:40:08,015] {{pod_launcher.py:148}} INFO - Wed Jun 23 12:40:08 UTC 2021
[2021-06-23 12:40:18,016] {{logging_mixin.py:104}} INFO - [2021-06-23 12:40:18,016] {{pod_launcher.py:148}} INFO - Wed Jun 23 12:40:18 UTC 2021
[2021-06-23 12:40:28,018] {{logging_mixin.py:104}} INFO - [2021-06-23 12:40:28,018] {{pod_launcher.py:148}} INFO - Wed Jun 23 12:40:28 UTC 2021
[2021-06-23 12:40:38,020] {{logging_mixin.py:104}} INFO - [2021-06-23 12:40:38,020] {{pod_launcher.py:148}} INFO - Wed Jun 23 12:40:38 UTC 2021
[2021-06-23 12:40:48,022] {{logging_mixin.py:104}} INFO - [2021-06-23 12:40:48,022] {{pod_launcher.py:148}} INFO - Wed Jun 23 12:40:48 UTC 2021
[2021-06-23 12:40:58,024] {{logging_mixin.py:104}} INFO - [2021-06-23 12:40:58,024] {{pod_launcher.py:148}} INFO - Wed Jun 23 12:40:58 UTC 2021
[2021-06-23 12:41:09,688] {{logging_mixin.py:104}} INFO - [2021-06-23 12:41:09,688] {{pod_launcher.py:154}} WARNING - Pod mwaa-pod-test.942c6c0987b14475a2a23cf250bb0c35 log read interrupted
[2021-06-23 12:41:09,774] {{logging_mixin.py:104}} INFO - [2021-06-23 12:41:09,774] {{pod_launcher.py:148}} INFO - Wed Jun 23 12:40:58 UTC 2021
[2021-06-23 12:41:09,796] {{logging_mixin.py:104}} INFO - [2021-06-23 12:41:09,796] {{pod_launcher.py:148}} INFO - Wed Jun 23 12:41:08 UTC 2021
[2021-06-23 12:41:18,028] {{logging_mixin.py:104}} INFO - [2021-06-23 12:41:18,028] {{pod_launcher.py:148}} INFO - Wed Jun 23 12:41:18 UTC 2021
[2021-06-23 12:41:28,029] {{logging_mixin.py:104}} INFO - [2021-06-23 12:41:28,029] {{pod_launcher.py:148}} INFO - Wed Jun 23 12:41:28 UTC 2021
[2021-06-23 12:41:38,031] {{logging_mixin.py:104}} INFO - [2021-06-23 12:41:38,031] {{pod_launcher.py:148}} INFO - Wed Jun 23 12:41:38 UTC 2021
[2021-06-23 12:41:48,033] {{logging_mixin.py:104}} INFO - [2021-06-23 12:41:48,033] {{pod_launcher.py:148}} INFO - Wed Jun 23 12:41:48 UTC 2021
[2021-06-23 12:41:58,035] {{logging_mixin.py:104}} INFO - [2021-06-23 12:41:58,035] {{pod_launcher.py:148}} INFO - Wed Jun 23 12:41:58 UTC 2021
```

Also, after 12 hours of running, i have these messages:
```
[2021-06-23 20:40:23,531] {{logging_mixin.py:104}} INFO - [2021-06-23 20:40:23,531] {{pod_launcher.py:148}} INFO - Wed Jun 23 20:40:23 UTC 2021
[2021-06-23 20:40:33,529] {{logging_mixin.py:104}} INFO - [2021-06-23 20:40:33,529] {{pod_launcher.py:148}} INFO - Wed Jun 23 20:40:33 UTC 2021
[2021-06-23 20:40:43,532] {{logging_mixin.py:104}} INFO - [2021-06-23 20:40:43,531] {{pod_launcher.py:148}} INFO - Wed Jun 23 20:40:43 UTC 2021
[2021-06-23 20:40:53,532] {{logging_mixin.py:104}} INFO - [2021-06-23 20:40:53,532] {{pod_launcher.py:148}} INFO - Wed Jun 23 20:40:53 UTC 2021
[2021-06-23 20:41:02,666] {{logging_mixin.py:104}} INFO - [2021-06-23 20:41:02,666] {{base_aws.py:368}} INFO - Airflow Connection: aws_conn_id=aws_default
[2021-06-23 20:41:02,754] {{logging_mixin.py:104}} INFO - [2021-06-23 20:41:02,754] {{base_aws.py:179}} INFO - No credentials retrieved from Connection
[2021-06-23 20:41:02,784] {{logging_mixin.py:104}} INFO - [2021-06-23 20:41:02,784] {{base_aws.py:87}} INFO - Creating session with aws_access_key_id=None region_name=eu-west-1
[2021-06-23 20:41:02,817] {{logging_mixin.py:104}} INFO - [2021-06-23 20:41:02,817] {{base_aws.py:157}} INFO - role_arn is None
[2021-06-23 20:41:02,876] {{taskinstance.py:871}} INFO - Dependencies not met for <TaskInstance: kubernetes_pod_example.pod-task 2021-06-23T08:41:01.157340+00:00 [running]>, dependency 'Task Instance State' FAILED: Task is in the 'running' state which is not a valid state for execution. The task must be cleared in order to be run.
[2021-06-23 20:41:02,949] {{taskinstance.py:871}} INFO - Dependencies not met for <TaskInstance: kubernetes_pod_example.pod-task 2021-06-23T08:41:01.157340+00:00 [running]>, dependency 'Task Instance Not Running' FAILED: Task is in the running state
[2021-06-23 20:41:02,974] {{logging_mixin.py:104}} INFO - [2021-06-23 20:41:02,974] {{local_task_job.py:93}} INFO - Task is not able to be run
[2021-06-23 20:41:03,534] {{logging_mixin.py:104}} INFO - [2021-06-23 20:41:03,534] {{pod_launcher.py:148}} INFO - Wed Jun 23 20:41:03 UTC 2021
[2021-06-23 20:41:15,223] {{logging_mixin.py:104}} INFO - [2021-06-23 20:41:15,223] {{pod_launcher.py:154}} WARNING - Pod mwaa-pod-test.942c6c0987b14475a2a23cf250bb0c35 log read interrupted
[2021-06-23 20:41:15,335] {{logging_mixin.py:104}} INFO - [2021-06-23 20:41:15,335] {{pod_launcher.py:148}} INFO - Wed Jun 23 20:41:03 UTC 2021
[2021-06-23 20:41:15,357] {{logging_mixin.py:104}} INFO - [2021-06-23 20:41:15,357] {{pod_launcher.py:148}} INFO - Wed Jun 23 20:41:13 UTC 2021
[2021-06-23 20:41:23,539] {{logging_mixin.py:104}} INFO - [2021-06-23 20:41:23,538] {{pod_launcher.py:148}} INFO - Wed Jun 23 20:41:23 UTC 2021
[2021-06-23 20:41:33,541] {{logging_mixin.py:104}} INFO - [2021-06-23 20:41:33,540] {{pod_launcher.py:148}} INFO - Wed Jun 23 20:41:33 UTC 2021
```
I'm not sure what to think about these messages, but i checked and Airflow did not attempt any update on the Kubernetes side (restarting / shooting down a task), meaning that it's only Airflow that is reacting maybe to a timeout parameter or so.
Also, I waited for 24h before killing the task, and these messages appear only once (compared to the 4h message that appears in cycle until the task is completed).

Hope it can bring some help to anyone !

**Comment by zappallot** [comment]
hi there,

we experience the same problem with Airflow 2.1.3 on Kubernetes 1.20.7, running as EKS service in AWS.
the sourcecode for detecting the problem was the following:
```
t1 = KubernetesPodOperator(
        task_id='bash-image',
        name='bash-image',
        image = 'bash:latest',
        cmds=[
            "bash",
            "-c",
            "while true; do echo 'hello bash'; sleep 10; done;"
            ],
        resources={'request_cpu': "1m", 'limit_cpu': "50m",
                   'request_memory': "8M", 'limit_memory': "32M"}
    )
```
the error message when this task fails:
```
[2021-10-28 09:48:12,961] {pod_launcher.py:149} INFO - hello bash
...
[2021-10-29 19:01:49,792] {pod_launcher.py:149} INFO - hello bash
[2021-10-29 19:01:52,639] {taskinstance.py:1462} ERROR - Task failed with exception
Traceback (most recent call last):
  File "/home/airflow/.local/lib/python3.8/site-packages/urllib3/response.py", line 697, in _update_chunk_length
    self.chunk_left = int(line, 16)
ValueError: invalid literal for int() with base 16: b''

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/airflow/.local/lib/python3.8/site-packages/urllib3/response.py", line 438, in _error_catcher
    yield
  File "/home/airflow/.local/lib/python3.8/site-packages/urllib3/response.py", line 764, in read_chunked
    self._update_chunk_length()
  File "/home/airflow/.local/lib/python3.8/site-packages/urllib3/response.py", line 701, in _update_chunk_length
    raise InvalidChunkLength(self, line)
urllib3.exceptions.InvalidChunkLength: InvalidChunkLength(got length b'', 0 bytes read)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/airflow/.local/lib/python3.8/site-packages/airflow/models/taskinstance.py", line 1164, in _run_raw_task
    self._prepare_and_execute_task_with_callbacks(context, task)
  File "/home/airflow/.local/lib/python3.8/site-packages/airflow/models/taskinstance.py", line 1282, in _prepare_and_execute_task_with_callbacks
    result = self._execute_task(context, task_copy)
  File "/home/airflow/.local/lib/python3.8/site-packages/airflow/models/taskinstance.py", line 1312, in _execute_task
    result = task_copy.execute(context=context)
  File "/home/airflow/.local/lib/python3.8/site-packages/airflow/providers/cncf/kubernetes/operators/kubernetes_pod.py", line 367, in execute
    final_state, remote_pod, result = self.create_new_pod_for_operator(labels, launcher)
  File "/home/airflow/.local/lib/python3.8/site-packages/airflow/providers/cncf/kubernetes/operators/kubernetes_pod.py", line 521, in create_new_pod_for_operator
    final_state, remote_pod, result = launcher.monitor_pod(pod=self.pod, get_logs=self.get_logs)
  File "/home/airflow/.local/lib/python3.8/site-packages/airflow/providers/cncf/kubernetes/utils/pod_launcher.py", line 147, in monitor_pod
    for line in logs:
  File "/home/airflow/.local/lib/python3.8/site-packages/urllib3/response.py", line 808, in __iter__
    for chunk in self.stream(decode_content=True):
  File "/home/airflow/.local/lib/python3.8/site-packages/urllib3/response.py", line 572, in stream
    for line in self.read_chunked(amt, decode_content=decode_content):
  File "/home/airflow/.local/lib/python3.8/site-packages/urllib3/response.py", line 793, in read_chunked
    self._original_response.close()
  File "/usr/local/lib/python3.8/contextlib.py", line 131, in __exit__
    self.gen.throw(type, value, traceback)
  File "/home/airflow/.local/lib/python3.8/site-packages/urllib3/response.py", line 455, in _error_catcher
    raise ProtocolError("Connection broken: %r" % e, e)
urllib3.exceptions.ProtocolError: ("Connection broken: InvalidChunkLength(got length b'', 0 bytes read)", InvalidChunkLength(got length b'', 0 bytes read))
[2021-10-29 19:01:52,642] {taskinstance.py:1505} INFO - Marking task as FAILED. dag_id=log_collection_tester-dv01, task_id=bash-image, execution_date=20211028T094738, start_date=20211028T094801, end_date=20211029T190152
```

During the same time when this error comes up, we detected **LeaderElection** events from four different Kubernetes control plane services in EKS, which are a blackbox for us.
- Events from the Lease kube-system/kube-scheduler (2021-10-29 19:01:58 +0000 UTC)
- Events from the Lease kube-system/kube-controller-manager (2021-10-29 19:01:59 +0000 UTC)
- Events from the ConfigMap kube-system/eks-certificates-controller (2021-10-29 19:02:11 +0000 UTC)
- Events from the ConfigMap kube-system/cp-vpc-resource-controller (2021-10-29 19:11:29 +0000 UTC)

These events take between 10 and 25 minutes and restarting processes during that time only result in another error like above.

So it looks to me like some maintenance on the Kubernetes control plane servers causes this problem on our Airflow system. The longer the task runs, the more likely it is that this error comes up in a process. 
This is quite an annoying error, since the time it takes to restart the control plane servers is so long, you cannot workaround that easily with a simple restart. We did catch this error in both TEST and PROD environments (we have two clusters).
Does anyone has an idea how to fix that or what a workaround could be without disabling the logging for all tasks?

**Comment by nikie** [comment]
This seems to be fixed by this PR: #17649,
available in Airflow 2.2: https://github.com/apache/airflow/blob/360474fff3738f70e95580a91e778250afa7ce82/airflow/providers/cncf/kubernetes/utils/pod_launcher.py#L258

**Comment by flrn77** [comment]
Hey there,

I can confirm that the error still exists in version 2.2.x python3.8.

For the setup I reference to @zappallot 2 comments above.
We have logging with datadog and frequently update process to new versions.
The errror occured in different k8sClusters and Versions 2.1.3, 2.1.4, 2.2.0, 2.2.1 and still present in 2.2.2.

**Comment by mikaeld** [comment]
@flrn77 what provider package version are you using for `cncf.kubernetes` when facing the issue?
The possible fix mentioned by @nikie for `cncf.kubernetes` version `2.0.3`, but you seem to mention Airflow "core" versions.

**Comment by flrn77** [comment]
Hi @mikaeld,
We have 2 clusters and 3 instances - latest setup I ran into it

cluster 1 (prod)
airflow-2.2.2
apache-airflow-providers-cncf-kubernetes 2.1.0 -> 5 times in the last month

cluster 2 (test)
airflow-2.2.3
apache-airflow-providers-cncf-kubernetes 2.2.0 - not noticed so far

cluster 2 (dev)
airflow-2.2.3
apache-airflow-providers-cncf-kubernetes 2.2.0 -> one time in the last month

**Comment by raphaelauv** [comment]
Since the latest version 3.0.0 , it's very stable , I have very long running jobs (more than 1 day ) , we can probably close this issue.

**Comment by ausiddiqui** [comment]
> Since the latest version 3.0.0 , it's very stable , I have very long running jobs (more than 1 day ) , we can probably close this issue.

@raphaelauv when you mention version 3.0.0 what are you referring to exactly? Airflow core or some other package/dependency?

**Comment by raphaelauv** [comment]
the version of the airflow cncf provider package ->

https://pypi.org/project/apache-airflow-providers-cncf-kubernetes/3.0.0/

**Comment by dstandish** [comment]
it appears this issue was resolved in PR https://github.com/apache/airflow/pull/17649

that PR was released in airflow 2.0.3

**Comment by ImNiceMan** [comment]
> ```python
> EarnestKubernetesPodOperator
> ```

where can I import EarnestKubernetesPodOperator?

ImportError: cannot import name 'EarnestKubernetesPodOperator' from 'airflow.providers.cncf.kubernetes.operators.kubernetes_pod' (/usr/local/lib/python3.7/site-packages/airflow/providers/cncf/kubernetes/operators/kubernetes_pod.py

**Comment by potiuk** [comment]
What? `EarnestKubernetesPodOperator` does not exist in Airflow Codebase. You must have some local modifications in your code.

**Comment by ImNiceMan** [comment]
> What? `EarnestKubernetesPodOperator` does not exist in Airflow Codebase. You must have some local modifications in your code.

Can you please give me a example?

**Comment by ImNiceMan** [comment]
> > What? `EarnestKubernetesPodOperator` does not exist in Airflow Codebase. You must have some local modifications in your code.
> 
> Can you please give me a example?

i dont know how to do ?

**Comment by potiuk** [comment]
> > > What? `EarnestKubernetesPodOperator` does not exist in Airflow Codebase. You must have some local modifications in your code.
> > 
> > 
> > Can you please give me a example?
> 
> i dont know how to do ?

Neither do I, sorry

**Comment by chelseatroy** [comment]
@ImNiceMan `Earnest` is the name of the company that Daniel (@dmateusp) works for. He made a subclass of `KubernetesPodOperator` to patch the issue, and he put the name of the company on the front of the name to indicate "this is a custom patch that we are using; not a part of the actuak Kubernetes library." 

So you can't import it, but you can make your own subclass in your code that fixes the issue the same way he did.

**Comment by chelseatroy** [comment]
The reason I'm here by the way folks is, this is totally happening again. I got this error on a job on June 25 and it has been failing ever since.

**Comment by potiuk** [comment]
> The reason I'm here by the way folks is, this is totally happening again. I got this error on a job on June 25 and it has been failing ever since.

I suggest you, if you know details of your Airlfow and reproduction steps, is to open a new issue. From what I see in the linked issue, you are using some external service and you have deveral issues and patches, without information of details of how to reproduce or even which versions of what you are using. There is 0 chance anyone will do anything on an issue closed for Airflow 2.0.3 which is likely even completely different issue that you have. Please open a new issue with all details and then there is a chance someone will help you.

**Comment by chelseatroy** [comment]
@potiuk I don't need help from the airflow team yet. I was just explaining why I'm responding to a conversation that's over a year old. 

In that conversation, someone on this issue asked for help with something, the context was all there to explain to him what was going on, and instead of doing that, you decided to play oblivious to make him feel bad. So I decided to offer him the explanation he needed the first time.

**Comment by potiuk** [comment]
> In that conversation, someone on this issue asked for help with something, and the context was all there to explain to them what was going on; in fact, I was able to figure it out by literally just reading the issue. Instead of helping them, you decided to play oblivious to make them feel bad. So I decided to offer them the explanation they needed the first time.

Your intentions are entirely unclear. I do not understand what you are trying to do. You are adding explanation and actually link the issue (by mentioning it in a random repository that is not at al connected to the original issue). The conversation has been closed in May 2022 and the question you are answering was asked also a year ago. And your answer brings no value whatsoever to no-one (including the one who asked it).

Just saying.
